qdb_ts.cpp

```diff
@@ -0,0 +1,331 @@
+#include "net_quasardb_qdb_jni_qdb.h"
+
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "helpers.h"
+#include <qdb/ts.h>
+
+void
+columns_to_native(JNIEnv * env, jobjectArray columns, qdb_ts_column_info * native_columns, size_t column_count) {
+  jfieldID name_field, type_field;
+  jclass object_class;
+  for (size_t i = 0; i < column_count; ++i) {
+    jobject object = (jobject) (env->GetObjectArrayElement(columns, i));
+
+    object_class = env->GetObjectClass(object);
+    name_field = env->GetFieldID(object_class, "name", "Ljava/lang/String;");
+    type_field = env->GetFieldID(object_class, "type", "I");
+    jstring name = (jstring)env->GetObjectField(object, name_field);
+
+    native_columns[i].type = (qdb_ts_column_type)(env->GetIntField(object, type_field));
+
+    // Is there a better way to do this ? Because we're using strdup here, we need a separate
+    // release function which is fragile.
+    native_columns[i].name = strdup(StringUTFChars(env, name));
```

Comment:
This is a fairly dubious approach, but if `release_native` is called appropriately, is stable. I couldn't directly see a more optimal way to handle this, but I might be missing something obvious.